File: objc/Project4/Project4/RemoteImageView.m
//
//  RemoteImageView.m
//  Project4
//
//  Created by Jinwoo Kim on 2/9/21.
//

#import "RemoteImageView.h"

@implementation RemoteImageView

- (NSURL *)getCachesDirectory {
    NSArray<NSURL *> *paths = [NSFileManager.defaultManager URLsForDirectory:NSCachesDirectory inDomains:NSUserDomainMask];
    return paths[0];
}

- (void)loadURL:(NSURL *)url {
    // stash the URL away for later checking
    self.url = url;
    
    // create a fate-to-save version of this URL that will be our cache filename
    // https://eddiekwon.github.io/swift/2018/09/01/Encoding101/
    NSString *savedFilename = [url.absoluteString stringByAddingPercentEncodingWithAllowedCharacters:NSCharacterSet.alphanumericCharacterSet];
    if (savedFilename == nil) return;
    
    // append that to the caches directory to get a complete path
    NSURL *fullPath = [[self getCachesDirectory] URLByAppendingPathComponent:savedFilename];
    
    // if the cached image exists already
    if ([NSFileManager.defaultManager fileExistsAtPath:fullPath.path]) {
        // use it and return
        self.image = [UIImage imageWithContentsOfFile:fullPath.path];
    }
    
    dispatch_async(dispatch_get_global_queue(QOS_CLASS_USER_INTERACTIVE, 0), ^{
        // download the image data
        NSData *imageData = [NSData dataWithContentsOfURL:url];
        if (imageData == nil) return;
        
        // write it to our cache file
        [imageData writeToURL:fullPath atomically:YES];
        
        // now the image has downloaded check it's still the one we want
        if (self.url == url) {
            dispatch_async(dispatch_get_main_queue(), ^{
                // update our image
                self.image = [UIImage imageWithData:imageData];
            });
        }
    });
}

@end

File: swift/Project4/Project4/RemoteImageView.swift
//
//  RemoteImageView.swift
//  Project4
//
//  Created by Jinwoo Kim on 2/8/21.
//

import UIKit

class RemoteImageView: UIImageView {
    var url: URL?
    
    func getCachesDirectory() -> URL {
        let paths = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)
        return paths[0]
    }
    
    func load(_ url: URL) {
        // stash the URL away for later checking
        self.url = url
        
        // create a safe-to-save version of this URL that will be our cache filename
        // https://eddiekwon.github.io/swift/2018/09/01/Encoding101/
        guard let savedFilename = url.absoluteString.addingPercentEncoding(withAllowedCharacters: CharacterSet.alphanumerics) else {
             return
        }
        
        // append that to the caches directory to get a complete path
        let fullPath = getCachesDirectory().appendingPathComponent(savedFilename)
        
        // if the cached image exists already
        if FileManager.default.fileExists(atPath: fullPath.path) {
            // use it and return
            image = UIImage(contentsOfFile: fullPath.path)
            return
        }
        
        // still here? Push work to a background thread
        DispatchQueue.global(qos: .userInteractive).async {
            // download the image data
            guard let imageData = try? Data(contentsOf: url) else { return }
            
            // write it to our cache file
            try? imageData.write(to: fullPath)
            
            // now the image has downloaded check it's still the one we want
            if self.url == url {
                DispatchQueue.main.async {
                    // update our image
                    self.image = UIImage(data: imageData)
                }
            }
        }
    }
}
